lumicks/pylake/kymo.py

```diff
@@ -261,39 +261,65 @@ def plot_with_time_histogram(self, color_channel, pixels_per_bin=1, hist_ratio=0
         ax_hist.set_ylabel("counts")
         ax_hist.set_title(self.name)
 
-    def downsampled_by(self, time_factor, reduce=np.sum):
+    def downsampled_by(
+        self, time_factor=1, position_factor=1, reduce=np.sum, reduce_timestamps=np.mean
+    ):
         """Return a copy of this Kymograph which is downsampled by `time_factor`
 
         Parameters
         ----------
         time_factor : int
-            The number of pixels that will be averaged in time.
+            The number of pixels that will be averaged in time (default: 1).
+        position_factor : int
+            The number of pixels that will be averaged in space (default: 1).
         reduce : callable
             The `numpy` function which is going to reduce multiple pixels into one.
             The default is `np.sum`.
+        reduce_timestamps : callable
+            The `numpy` function that is going to reduce timestamps into one.
+            The default is `np.mean`.
         """
         result = Kymo(self.name, self.file, self.start, self.stop, self._json)
 
         def image_factory(_, channel):
             data = self._image(channel)
-            return block_reduce(data, (1, time_factor), func=reduce)[
-                :, : data.shape[1] // time_factor
+            return block_reduce(data, (position_factor, time_factor), func=reduce)[
+                : data.shape[0] // position_factor, : data.shape[1] // time_factor
```

Comment:
I am trying to make sense of the data slicing here, but I am failing to parse it.  `data[ : foo, : bar ]` looks very strange to the untrained eye.  Can you perhaps explain it to me?

Reply:
It's a two dimensional matrix. `: foo` just means up to foo for that axis. We do this for both axes.

What we do here is slice to the nearest integer multiple of the axis dimension. So let's say the original had the last pixel only half filled (not ending on a full multiple of the downsampling factor) then we drop that sample in the resulting array.

Let's say we had a 5 by 5 matrix `(5, 5)`.
Then `data.shape[0] // 2` would be 2. Hence we'd only end up including row 0 and 1 and drop the last one.